3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,366 @@
+            case JArray(contexts) =>
+              JArray(contexts.map {
+                case JObject(context) => modifyObjectIfSchemaMatches(context)
+                case x                => x
+              })
+            case JObject(unstructEvent) => modifyObjectIfSchemaMatches(unstructEvent)
+            case x                      => x
+          }
+          JObject((jObjectMap ++ updated).toList)
+        }
+        case x => x
+      }))
+    } else null
+
+  private def modifyObjectIfSchemaMatches(context: List[(String, json4s.JValue)]): JObject = {
+    val fieldsObj = context.toMap
+    (for {
+      schema              <- fieldsObj.get("schema")
+      parsedSchemaMatches <- SchemaKey.parse(schema.extract[String]).map(schemaCriterion.matches).toOption
+      data                <- fieldsObj.get("data")
+      if parsedSchemaMatches
+    } yield JObject(fieldsObj.updated("schema", schema).updated("data", jsonPathReplace(data)).toList)).getOrElse(JObject(context))
+  }
+
+  // Configuration for JsonPath
+  private val JacksonNodeJsonObjectMapper = {
+    val objectMapper = new ObjectMapper()
+    objectMapper.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false)
+    objectMapper
+  }
+  private val JsonPathConf =
+    Configuration
+      .builder()
+      .options(JOption.SUPPRESS_EXCEPTIONS)
+      .jsonProvider(new JacksonJsonNodeJsonProvider(JacksonNodeJsonObjectMapper))
+      .build()
+
+  /**
+   * Replaces a value in the given context data with the result of applying the strategy that value.
+   */
+  private def jsonPathReplace(jValue: JValue): JValue = {
+    val objectNode      = JsonMethods.mapper.valueToTree[ObjectNode](jValue)
+    val documentContext = JJsonPath.using(JsonPathConf).parse(objectNode)
+    documentContext.map(
+      jsonPath,
+      ScrambleMapFunction(strategy)
+    )
+    JsonMethods.fromJsonNode(documentContext.json[JsonNode]())
+  }
+}
+
+case class ScrambleMapFunction(val strategy: PiiStrategy) extends MapFunction {
+  override def map(currentValue: AnyRef, configuration: Configuration): AnyRef = currentValue match {
+    case s: String => strategy.scramble(s)
+    case a: ArrayNode =>
+      a.elements.asScala.map {
+        case t: TextNode     => strategy.scramble(t.asText())
+        case default: AnyRef => default
+      }
+    case default: AnyRef => default
+  }
+}
+
+/**
+ * Implements a pseudonymization strategy using any algorithm known to MessageDigest
+ * @param hashFunction the MessageDigest function to apply
+ */
+case class PiiStrategyPseudonymize(hashFunction: MessageDigest) extends PiiStrategy {
+  val TextEncoding                                 = "UTF-8"
+  override def scramble(clearText: String): String = hash(clearText)
+  def hash(text: String): String = {
+    val bytes = hashFunction.digest(text.getBytes(TextEncoding))
+    String.format("%0" + bytes.size*2 + "x", new java.math.BigInteger(1, bytes))
```

Comment:
@chuwy What do you think about this formatting string?

Reply:
Yes, it looks good to me. Though I don't think I'm aware of all implications it has.